### PR TITLE
29 use svelte stores

### LINKED ISSUE #29: use Svelte stores

### PR DESCRIPTION
**Pull request by Rettend**
- restructured the mess in the `src` folder (like component files being in `routes` :D )
- it turned out that it's fine if a component modifies its props if the prop is a `$bindable` prop:
  - we just needed two lines 🤦
- created the items store in `stores/item.svelte.ts`, and it encapsulates its methods
  - the svelte.ts extension is just for using svelte runes in ts files
  - frankly, extracting all that logic to a store wouldn't have been necessary, but i learnt how stores work and this allows accessing the items from anywhere in the entire app, like we may want to do something with the selected item somewhere else idk. and it does look cleaner imo
- simplified some stuff, and i now know what the problem with the Item tests was

### CODE REVIEW

File: src/lib/components/StoragePlace.svelte [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+<div
+  class="rounded-sm border m-3 inline-block h-fit min-w-80 max-w-[420px] border-black p-1"
+  role="tree"
+>
+  <h1 class="font-bold">{storagePlaceName} <span class="text-stone-400">({items.list.length})</span></h1>
+  <div role="group">
+    {#each items.list as item, i (item.id)}
+      <Item
+        bind:item={items.list[i]}
```
Comment (Rettend) [inline src/lib/components/StoragePlace.svelte:33]: the parent component binds to the child's bindable `item` prop

File: src/lib/components/item/Item.svelte [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+
+  export type Events = {
+    onSelected: (index?: number) => void
+    onDelete: (itemId: string) => void
+  }
+</script>
+
+<script lang="ts">
+  const {
+    item = $bindable(),
```
Comment (Rettend) [inline src/lib/components/item/Item.svelte:21]: the item prop is bindable by the parent

File: src/lib/components/item/ItemWrappedInTree.svelte [hunk 3]
```diff
@@ -1,16 +1,16 @@
...
 -->
 
 <script lang="ts">
-	import Item from '../routes/Item.svelte';
-	import type { Props, Events } from '../routes/Item.svelte';
+  import Item from './Item.svelte'
+  import type { Events, Props } from './Item.svelte'
 
-	let { ...props }: Props & Events = $props();
+  const { ...rest }: Props & Events = $props()
```
Comment (Rettend) [inline src/lib/components/item/ItemWrappedInTree.svelte:11]: the name `props` conflicted with the `$props` rune's name wtf
so it's now called `rest`

File: src/lib/components/item/itemGenerator.ts [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    shelfLife: 7,
+  },
+  {
+    id: uuid(),
+    name: 'cauliflower',
+    quantity: 1,
+    dateAdded: dayjs().subtract(4, 'day').format('YYYY-MM-DD'),
+    shelfLife: 7,
+  },
+] satisfies StoredItem[]
```
Comment (Rettend) [inline src/lib/components/item/itemGenerator.ts:167]: finally i can use the new satisfies typescript keyword!!!

File: src/lib/stores/item.svelte.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
+import { v7 as uuid } from 'uuid'
+import dayjs from 'dayjs'
+import { possibleItems } from '$lib/components/item/itemGenerator'
+import { randomIntFromInterval } from '$lib/utils'
+import type { StoredItem } from '$lib/types'
+
+export function createItemStore() {
```
Comment (Rettend) [inline src/lib/stores/item.svelte.ts:7]: this is how a store looks.
we use getters to not lose reactivity (the docs say destructuring loses reactivity)

File: src/routes/+page.svelte [hunk 6]
```diff
@@ -1,77 +1,62 @@
...
+  import StoragePlace from '$lib/components/StoragePlace.svelte'
 
-	let showTips = $state(false);
-	// let tips: HTMLDivElement | null = $state(null);
-	let tips: HTMLDivElement;
+  let showTips = $state(false)
+  let tips: HTMLDivElement | null = $state(null)
 
-	// TODO: fix broken accordion on resize
+  $effect(() => {
```
Comment (Rettend) [inline src/routes/+page.svelte:7]: just using `let tips: HTMLDivElement` gave a warning so to use `$state` on an element, we modify it in an `$effect` (and not in the template).
here, `$effect` acts as the old `onMount`, so runs once during component initialization.

File: tailwind.config.js [hunk 7]
```diff
@@ -1,13 +1,32 @@
...
+  theme: {
+    extend: {
+      transitionProperty: {
+        height: 'height, max-height',
+        margin: 'margin',
+      },
+    },
+  },
+  plugins: [
+    addIconSelectors({
```
Comment (Rettend) [inline tailwind.config.js:15]: this setup is for iconify when using tailwind.
and i made the icons bolder to match how they were displayed previously

File: src/lib/components/item/itemGenerator.ts (hunk 4 above)
Comment (ashlsun) [reply src/lib/components/item/itemGenerator.ts:167]: 😮 this is new to me

File: static/Hershey-Noailles-Futura-Duplex-Regular.ttf
Comment (ashlsun) [inline static/Hershey-Noailles-Futura-Duplex-Regular.ttf:1]: so if this doesn't exist anymore, should we remove the url to the ttf files in app.css?

**Review by ashlsun — APPROVED** [review]
thanks for your work cleaning things up! 
`bun i` was significantly faster than `npm i` on my mac  :D

also, what exactly was the problem you discovered with Item tests?!

File: static/Hershey-Noailles-Futura-Duplex-Regular.ttf
Comment (Rettend) [reply static/Hershey-Noailles-Futura-Duplex-Regular.ttf:1]: yes, i didn't notice those. woff alone is fine, no need for both. i'll make a fix pr

**Comment by Rettend** [comment]
> 😮 this is new to me

okay here we go if you haven't looked it up yet:

```ts
// we have an object type
type User = {
  id: number
  name: string
}

// user is of type User
const user: User = {
  id: 1,
  name: 'asd'
}

// user's type is { id: number, name: string }
const user = {
  id: 1,
  name: 'asd'
} satisfies User
```

so `satisfies` will keep the inferred type, but will still provide typechecking for user.

In our case, both would work, but there are cases where you have to use `satisfies`

**Comment by Rettend** [comment]
> thanks for your work cleaning things up! 
> `bun i` was significantly faster than `npm i` on my mac  :D
> 
> also, what exactly was the problem you discovered with Item tests?! 

because of how we used events like deleteItem, when the quantity was 1 and you pressed the left arrow, it would not decrease it (because it would return early). so i increased the quantity to 2, and made two tests, one where it should decrease quantity and one where it should call the delete event.

i'm not sure how to write this clearer

**Pull request merged**